tests/foreman/ui/test_reporttemplates.py

```diff
@@ -396,23 +149,112 @@ def test_positive_e2e_crud(session, module_org, module_loc):
+                query={'search': 'name={0}'.format(name)})[0].read()
+        assert rt.name == name
+        assert rt.organization[0].read().name == module_org.name
+        assert rt.location[0].read().name == module_loc.name
+        assert rt.template == content
+        assert rt.default is True
+        # READ report template
+        rt = session.reporttemplate.read(name)
+        assert rt['template']['name'] == name
+        assert rt['template']['default'] is True
+        assert rt['template']['template_editor']['editor'] == content
+        assert rt['inputs'][0]['name'] == input_name
+        assert rt['inputs'][0]['required'] is True
+        assert rt['inputs'][0]['input_type'] == 'User input'
+        assert rt['type']['snippet'] is False
+        assert rt['locations']['resources']['assigned'][0] == module_loc.name
+        assert rt['organizations']['resources']['assigned'][0] == module_org.name
+        # UPDATE report template
         session.reporttemplate.update(
             name,
             {'template.name': new_name, 'type.snippet': True}
         )
+        # check this by API once BZ1702676 gets fixed
+        # assert entities.ReportTemplate(name=new_name).search(
+        #    query={'search': 'name={0}'.format(new_name)})[0].read().snippet is True
         assert session.reporttemplate.search(new_name)[0]['Name'] == new_name
-        updated_pt = session.reporttemplate.read(new_name)
-        assert updated_pt['type']['snippet'] is True
+        rt = session.reporttemplate.read(new_name)
+        assert rt['type']['snippet'] is True
+        # LOCK
+        session.reporttemplate.lock(new_name)
+        # check this by API once BZ1702676 gets fixed
+        # assert rt.locked is True
+        assert session.reporttemplate.is_locked(new_name) is True
+        # UNLOCK
+        session.reporttemplate.unlock(new_name)
+        # check this by API once BZ1702676 gets fixed
+        # assert rt.locked is False
+        assert session.reporttemplate.is_locked(new_name) is False
+        # CLONE
+        session.reporttemplate.clone(
+            new_name,
+            {
+                'template.name': clone_name,
+            }
+        )
+        rt = entities.ReportTemplate().search(
+                query={'search': 'name={0}'.format(clone_name)})[0].read()
+        assert rt.name == clone_name
+        assert rt.organization[0].read().name == module_org.name
+        assert rt.location[0].read().name == module_loc.name
+        assert rt.template == content
+        assert rt.default is True
+        # EXPORT
+        file_path = session.reporttemplate.export(new_name)
+        assert os.path.isfile(file_path)
+        with open(file_path) as dfile:
+            assert content in dfile.read()
+        # DELETE report template
         session.reporttemplate.delete(new_name)
-        assert not session.reporttemplate.search(new_name)
+        assert len(entities.ReportTemplate().search(
```

Comment:
once more, if read is working then search is also working

Reply:
This tests that the report template is deleted

Reply:
right, so verify it through UI. I can't imagine scenario where API validation is more preferable

Reply:
Whenever there is discrepancy between API and WebUI. Testing through API can potentially find more bugs and is also more efficient.

Reply:
Right, and that is a reason we have three different folders per each type. Just to test API in API folder, CLI in CLI folder and UI in UI folder. In that PR we have UI folder, so test through UI

Reply:
In this part, I am testing WebUI by running tasks in WebUI and checking the results through API. That doesn't mean I am testing API, that means I am testing WebUI and I am not relying on the WebUI while doing so (which makes sense since I am **testing** it).

Reply:
In this part, I am testing WebUI by running tasks in WebUI and checking the results through API. That doesn't mean I am testing API, that means I am testing WebUI and I am not relying on the WebUI while doing so (which makes sense since I am **testing** it).

Reply:
API is far faster. Only risk using API instead of UI to verify the entity was deleted is if UI would still be showing the entity after it was deleted. I do not recall any issue like that, so IMO speed benefit overweights this risk.

Reply:
hmm. The whole idea of test automation is to verify result through same interface it was created.

Example:
You created through UI, deleted through UI, validated successfully through API (result showed entity is not present in application), but UI is still showing entity. So your test proved nothing.

We never, ever validate through different interfaces. Yes, we use API to speed up pre-configuration procedures, but attempt to use it for verifying results ruin the whole ideology of test automation. I can't even explain how many fundamental principles of quality assurance is violated by proposed idea.

Let's not waste time anymore on that pointless discussion

Reply:
Imagine a simple example, where there are two databases - WebUI only uses one and API only uses the second, so whatever you do in WebUI is never visible through API. Which is obviously wrong. Yet, using WebUI to verify WebUI action results would never catch this.

Reply:
> Let's not waste time anymore on that pointless discussion

OK, lets ack the PR then :-)

Reply:
Not a chance =)

Reply:
interesting example, but we do not test that frequently. For such scenario, we always create separate, independent test case and usually for existing issue as it will be too redundant to add so many test cases.

Reply:
If you feel that it is must have validation for you, then validate whether both UI and API or only UI. I can't ack only API validation